File: fsharp/yacht/Yacht.fs
module Yacht

type Category = 
    | Ones
    | Twos
    | Threes
    | Fours
    | Fives
    | Sixes
    | FullHouse
    | FourOfAKind
    | LittleStraight
    | BigStraight
    | Choice
    | Yacht

type Die =
    | One 
    | Two 
    | Three
    | Four 
    | Five 
    | Six

let score (category: Category) (dice: Die list) =
    let countDice die =
        dice
        |> List.filter (fun x -> x = die)
        |> List.length
        
    let freqs =
        [Die.One; Die.Two; Die.Three; Die.Four; Die.Five; Die.Six]
        |> List.map countDice
        
    let number die =
        match die with
        | Die.One -> 1
        | Die.Two -> 2
        | Die.Three -> 3
        | Die.Four -> 4
        | Die.Five -> 5
        | Die.Six -> 6
        
    let numbers dice = dice |> List.map number
    
    let sumDice dice = dice |> numbers |> List.sum
        
    match category with
    | Ones  -> countDice Die.One
    | Twos  -> 2 * countDice Die.Two
    | Threes -> 3 * countDice Die.Three
    | Fours -> 4 * countDice Die.Four
    | Fives -> 5 * countDice Die.Five
    | Sixes -> 6 * countDice Die.Six
    | FullHouse ->
        match (List.sort freqs) with
        | [_; _; _; _; 2; 3] -> sumDice dice
        | _ -> 0
    | FourOfAKind ->
        let topFreq = List.max freqs
        match topFreq with
        |4 | 5 -> 4 * ((List.findIndex (fun x -> x = topFreq) freqs) + 1)
        | _ -> 0
    | LittleStraight ->
        match freqs with
        | [1; 1; 1; 1; 1; _] -> 30
        | _ -> 0
    | BigStraight ->
        match freqs with
        | [_; 1; 1; 1; 1; 1] -> 30
        | _ -> 0
    | Choice -> sumDice dice
    | Yacht -> if (List.max freqs) = 5 then 50 else 0
     

File: python/yacht/yacht.py
# Score categories.
# Change the values as you see fit.
YACHT = 0
ONES = 1
TWOS = 2
THREES = 3
FOURS = 4
FIVES = 5
SIXES = 6
FULL_HOUSE = 7
FOUR_OF_A_KIND = 8
LITTLE_STRAIGHT = 9
BIG_STRAIGHT = 10
CHOICE = 11


def score(dice, category):
    if category == YACHT:
        return 50 if max(freqs(dice)) == 5 else 0
    if category in [ONES, TWOS, THREES, FOURS, FIVES, SIXES]:
        return sum(x for x in dice if x == category)
    if category == FULL_HOUSE:
        return sum(dice) if sorted(freqs(dice)) == [0, 0, 0, 0, 2, 3] else 0
    if category == FOUR_OF_A_KIND:
        freq_list = freqs(dice)
        maxindex = max(freq_list)
        return 4 * (freq_list.index(max(freq_list)) + 1) if maxindex >= 4 else 0
    if category == LITTLE_STRAIGHT:
        return 30 if sorted(dice) == [1, 2, 3, 4, 5] else 0
    if category == BIG_STRAIGHT:
        return 30 if sorted(dice) == [2, 3, 4, 5, 6] else 0
    if category == CHOICE:
        return sum(dice)


def freqs(dice):
    return [len([x for x in dice if x == d]) for d in range(1, 7)]
